### ISSUE TITLE
custom runner cannot be tested with the current documentation setup

### ISSUE THREAD

**Issue by iQuxLE** [monarch-initiative/pheval#269]
This issue started with having a problem with the pydantic version as `poetry add pheval` was not working correctly. So I still want to make sure this is still covered, even when `poetry add git+https://github.com/monarch-initiative/pheval.git` solved pydantic (https://github.com/monarch-initiative/pheval/issues/266).

I wanted to check:
`
pheval run -i input.txt -t './test' -r 'exoragphevalrunner' -o out.txt
`

on this Runner:
```
"""Custom Pheval Runner."""
import json
from dataclasses import dataclass
from pathlib import Path
from pheval.runners.runner import PhEvalRunner


@dataclass
class ExoRagPhevalRunner(PhEvalRunner):

    """CustomPhevalRunner Class."""

    input_dir: Path  #
    testdata_dir: Path
    tmp_dir: Path
    output_dir: Path
    config_file: Path
    version: str


    def prepare(self):
        """Prepare method."""
        print("preparing")

    def run(self):
        """Run method."""
        print("running with custom pheval Runner")

    def post_process(self):
        """Post-process method."""
        print("No results to process")


```
I did this to update poetry:
```
❯ poetry install
Installing dependencies from lock file
No dependencies to install or update
Installing the current project: exomiser_rag_embedding_runner (0.0.0)

```
Than:
```
❯ pheval run -i input.txt -t './test' -r 'exoragphevalrunner' -o out.txt

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/bin/pheval", line 5, in <module>
    from pheval.cli import pheval
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pheval/cli.py", line 9, in <module>
    from .cli_pheval_utils import (
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pheval/cli_pheval_utils.py", line 7, in <module>
    from pheval.prepare.create_noisy_phenopackets import scramble_phenopackets
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pheval/prepare/create_noisy_phenopackets.py", line 4, in <module>
    from oaklib.implementations.pronto.pronto_implementation import ProntoImplementation
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/oaklib/__init__.py", line 9, in <module>
    from oaklib.selector import get_implementation_from_shorthand  # noqa:F401
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/oaklib/selector.py", line 8, in <module>
    from oaklib.implementations import GildaImplementation
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/oaklib/implementations/__init__.py", line 5, in <module>
    from oaklib.implementations.funowl.funowl_implementation import FunOwlImplementation
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/oaklib/implementations/funowl/funowl_implementation.py", line 28, in <module>
    from oaklib.interfaces.patcher_interface import PatcherInterface
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/oaklib/interfaces/patcher_interface.py", line 21, in <module>
    from oaklib.utilities.kgcl_utilities import generate_change_id
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/oaklib/utilities/kgcl_utilities.py", line 11, in <module>
    import kgcl_schema.grammar.parser as kgcl_parser
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/kgcl_schema/grammar/parser.py", line 9, in <module>
    from bioregistry import parse_iri, get_preferred_prefix, curie_to_str
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/bioregistry/__init__.py", line 5, in <module>
    from .collection_api import get_collection, get_context  # noqa:F401
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/bioregistry/collection_api.py", line 7, in <module>
    from .resource_manager import manager
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/bioregistry/resource_manager.py", line 36, in <module>
    from .schema import (
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/bioregistry/schema/__init__.py", line 5, in <module>
    from .struct import (  # noqa:F401
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/bioregistry/schema/struct.py", line 35, in <module>
    from bioregistry.utils import curie_to_str, deduplicate, removeprefix, removesuffix
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/bioregistry/utils.py", line 25, in <module>
    from pydantic.json import ENCODERS_BY_TYPE
ImportError: cannot import name 'ENCODERS_BY_TYPE' from 'pydantic.json' (/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pydantic/json.py)
```

I was expecting:
```
preparing
running with custom pheval Runner
No results to process

```

I have this version of pydantic after adding pheval to poetry

```
❯ pip show pydantic

Name: pydantic
Version: 2.5.3
Summary: Data validation using Python type hints
Home-page: 
Author: 
Author-email: Samuel Colvin <s@muelcolvin.com>, Eric Jolibois <em.jolibois@gmail.com>, Hasan Ramezani <hasan.r67@gmail.com>, Adrian Garcia Badaracco <1755071+adriangb@users.noreply.github.com>, Terrence Dorsey <terry@pydantic.dev>, David Montague <david@pydantic.dev>, Serge Matveenko <lig@countzero.co>, Marcelo Trylesinski <marcelotryle@gmail.com>, Sydney Runkle <sydneymarierunkle@gmail.com>, David Hewitt <mail@davidhewitt.io>
License: 
Location: /Users/carlo/Library/Caches/pypoetry/virtualenvs/exomiser-rag-embedding-runner-rrKu6h-f-py3.11/lib/python3.11/site-packages
Requires: annotated-types, pydantic-core, typing-extensions
Required-by: bioregistry, linkml
```

After I added:
`poetry add git+https://github.com/monarch-initiative/pheval.git`

```
❯ pheval run -i input.txt -t './test' -r 'ExoRagPhevalRunner' -o out.txt

Traceback (most recent call last):
  File "/Users/carlo/Library/Caches/pypoetry/virtualenvs/exomiser-rag-embedding-runner-rrKu6h-f-py3.11/bin/pheval", line 8, in <module>
    sys.exit(pheval())
             ^^^^^^^^
  File "/Users/carlo/Library/Caches/pypoetry/virtualenvs/exomiser-rag-embedding-runner-rrKu6h-f-py3.11/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/carlo/Library/Caches/pypoetry/virtualenvs/exomiser-rag-embedding-runner-rrKu6h-f-py3.11/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/carlo/Library/Caches/pypoetry/virtualenvs/exomiser-rag-embedding-runner-rrKu6h-f-py3.11/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/carlo/Library/Caches/pypoetry/virtualenvs/exomiser-rag-embedding-runner-rrKu6h-f-py3.11/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/carlo/Library/Caches/pypoetry/virtualenvs/exomiser-rag-embedding-runner-rrKu6h-f-py3.11/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/carlo/Library/Caches/pypoetry/virtualenvs/exomiser-rag-embedding-runner-rrKu6h-f-py3.11/lib/python3.11/site-packages/pheval/cli_pheval.py", line 86, in run
    runner_class = get_implementation_resolver().lookup(runner)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/carlo/Library/Caches/pypoetry/virtualenvs/exomiser-rag-embedding-runner-rrKu6h-f-py3.11/lib/python3.11/site-packages/class_resolver/api.py", line 176, in lookup
    return get_cls(
           ^^^^^^^^
  File "/Users/carlo/Library/Caches/pypoetry/virtualenvs/exomiser-rag-embedding-runner-rrKu6h-f-py3.11/lib/python3.11/site-packages/class_resolver/api.py", line 378, in get_cls
    raise KeyError(
KeyError: "Invalid PhEvalRunner name: ExoRagPhevalRunner (normalized to: exoragphevalrunner). Valid choices are: ['defaultphevalrunner']"
```

So what I had to do is using 'defaultphevalrunner', add a config.yaml in the root dir and a output dir:
and change the command to:

` pheval run -i . -t './test' -r 'defaultphevalrunner' -o output/
`
To get:
```
preparing
running
post processing
```

But this is not coming from my implemented custom runner.

However, whats not written in the documentation and also not added automatically by the cookiecutter template is the `config.yaml` as `-i` is looking for a directory with a config.yaml. Same goes for the output dir. `-o` looks for an output directory not a file, which also should have been created by the template in the beginning.

However I think still the biggest thing I could not figure out is how to actually run my `exoragphevalrunner` and not only the `defaultphevalrunner`.

And the `config.yaml` should be added with this input I think, either to the documentation or by cookiecutter:

```
tool: exampleTool
tool_version: 1
variant_analysis: False
gene_analysis: False
disease_analysis: True
tool_specific_configuration_options:
```

currently the only thing thats getting created is a `.pre-commit-config.yaml`:
```
default_language_version:
  python: python3
repos:
- repo: https://github.com/pre-commit/pre-commit-hooks
  rev: v4.4.0
  hooks:
  - id: end-of-file-fixer
  - id: check-yaml
  - id: check-toml
  - id: trailing-whitespace
- repo: https://github.com/psf/black
  rev: 23.7.0
  hooks:
  - id: black
- repo: https://github.com/astral-sh/ruff-pre-commit
  rev: v0.0.278
  hooks:
  - id: ruff
    args: [--fix, --exit-non-zero-on-fix]
- repo: https://github.com/codespell-project/codespell
  rev: v2.2.4
  hooks:
  - id: codespell
    additional_dependencies:
      - tomli
```

/Carlo

Edit: 
I also tried using the 
```
// current runner implementation ...
   def prepare(self): 
   ...
   def run(self):
   ...
   def postprocess(self):
  ...

    def construct_meta_data(self):
        """Add metadata."""
        self.meta_data.tool_specific_configuration_options = CustomisedMetaData(customised_field="customised_value")
        return self.meta_data
```
and a `CustomisedMetaData` class as that was in the documentation but it did not change the given problem:
```
from dataclasses import dataclass

@dataclass
class CustomisedMetaData:
    customised_field: str

```

**Comment by yaseminbridges** [comment]
Ah yes. I had this error before - sometimes it comes up for me when creating a runner but recently it has not. From what I understand there are breaking changes in the latest pydantic release. For now, if you just run `pip install pydantic==1.10.13` it should run the PhEval run command. I will have to work on getting the latest pydantic working in PhEval

**Comment by yaseminbridges** [comment]
The `defaultphevalrunner` is just the base class we have in the main PhEval setup to make sure that the base runner is runs as we expect. It is where we call the abstract methods, `prepare`, `run` and `post_process` for override in implementations as well as setting up the directory structure in the output. 

The `config.yaml` that needs to be placed within the input directory is outlined in the documentation:

https://monarch-initiative.github.io/pheval/developing_a_pheval_plugin/#additional-tool-specific-configurations

This is not part of the cookie-cutter template as it does not need to be within the project and the cookie-cutter template is for Monarch projects, not just the PhEval runners. The idea is that you may have the input directory located somewhere separate from the project and this is what you provide on the CLI. In other runner projects for example the `pheval.exomiser` the `config.yaml` is just placed in the project as an example of what needs to be placed in the input directory.

The output directory is also separate from the project structure.

**Comment by iQuxLE** [comment]
> The `defaultphevalrunner` is just the base class we have in the main PhEval setup to make sure that the base runner is runs as we expect. It is where we call the abstract methods, `prepare`, `run` and `post_process` for override in implementations as well as setting up the directory structure in the output.
> 
> The `config.yaml` that needs to be placed within the input directory is outlined in the documentation:
> 
> https://monarch-initiative.github.io/pheval/developing_a_pheval_plugin/#additional-tool-specific-configurations
> 
> This is not part of the cookie-cutter template as it does not need to be within the project and the cookie-cutter template is for Monarch projects, not just the PhEval runners. The idea is that you may have the input directory located somewhere separate from the project and this is what you provide on the CLI. In other runner projects for example the `pheval.exomiser` the `config.yaml` is just placed in the project as an example of what needs to be placed in the input directory.
> 
> The output directory is also separate from the project structure.

ok I missed that config.yaml 

but what about testing the '`customrunner`'? I thought thats intended when using this test command.
Or is it only for the BaseClass default runner?

**Comment by yaseminbridges** [comment]
> but what about testing the 'customrunner'? I thought thats intended when using this test command.
Or is it only for the BaseClass default runner?

I'm a bit confused by the question? Do you mean testing the 'customrunner' with the `pheval run` command? If so, then yes running `pheval run -r exoragphevalrunner ...` will run all code you have outlined in your `ExoRagPhevalRunner` class and in a way test it all works. It failed due to the problem with the `pydantic` version
